### ISSUE TITLE
"make test" and got the error "no matching function for call to ReadImageToDatum and ReadFileToDatum"

### ISSUE THREAD

**Issue by liuyuying0829** [lhmRyan/deep-supervised-hashing-DSH#7]
Dear authors,
when i follow the instructions on "http://caffe.berkeleyvision.org/installation.html" to compile the source code, when i set "USE_CUDNN := 1",i got the error: few arguements for cudnn::createPoolingDesc ,i compared the same file with caffe and copy the caffe's code and then compile this file successful. But when i continued compile other files also have this error and then i known these problems are caused by the version of CuDnn . So i set"#USE_CUDNN := 1",then i can "make pycaffe" and "make all" successful ,but when i want to "make test",i got the error "no matching function for call to ReadImageToDatum and ReadFileToDatum",i also compiled this file with caffe's and they are the same so i can't resolve this problem, can you give me some advise ?

**Comment by lhmRyan** [comment]
about cuDNN: Our codes use cudnn v4. You can download the tar file of cudnn v4, untar it, and add the "include" and "lib64" directories to Makefile.config (e.g. INCLUDE_DIRS := $(PYTHON_INCLUDE) /usr/local/include /home/liu/cudnnv4/include; LIBRARY_DIRS := $(PYTHON_LIB) /usr/local/lib /usr/lib /home/liu/cudnnv4/lib64), and then you should be able to use compile with cudnn.

about "make test": I modified the code of "io.cpp" to support multi-label LMDB/leveldb, but I didn't modify the codes for testing, thus "make test" won't work here. However, our codes works normally as long as you can run "make all" successfully.

**Comment by liuyuying0829** [comment]
Thank you so much !!! i used "make test -i" to ignored the errors(crazy...)and follow your example usage CIFAR-10/train_full.sh to continue  your experiment and i did't use the cudnn and now it running ,but the speed is not fast,about 12 seconds per 100 iterations ,the loss is ave 1.2,is that normal?

**Comment by lhmRyan** [comment]
Without cudnn, 12 seconds per 100 iterations is normal. With our codes, the final loss on CIFAR-10 is about 0.6 (12-bit code), and the corresponding retrieval mAP is about 0.67.

**Comment by liuyuying0829** [comment]
Thank you for your help!!!

**Comment by liuyuying0829** [comment]
Dear authors,
l have a question about why you use the image pair as the training input ?

**Comment by lhmRyan** [comment]
We use the image pairs and their corresponding similarity labels so that the network could learn to preserve the relationships of images defined by the similarity labels. Such scheme is widely adopted in many hashing methods. 

Actually, other loss functions for metric learning, triplet ranking loss for example, could also be adopted, which we've evaluated in our recent experiments. Therefore, the use of image pair is just an option, while other options are also available.

**Comment by liuyuying0829** [comment]
Thank you so much for your detailed and patient explanation!! If i may ask? Do you have personal homepage ?Because i am very interested in your research and want to learn from your study.

**Comment by lhmRyan** [comment]
My personal homepage is still under construction. If you have any question, feel free to contact me via email.

**Comment by ahhan02** [comment]
Dear lhmRyan, I followed your advices to include cudnnv4, and run "make all" successfully. 

However, my model didn't converge at all and the loss is 10.8 avg. b.t.w., I haven't changed any code except Makefile.config!!!  

I have no idea about it, could you give me some advice?? @lhmRyan

**Comment by lhmRyan** [comment]
@ahhan02 It's strange. I just cloned this repository to my computer, compiled it, downloaded the data of cifar-10, and trained the model. After 100 iterations, the loss drops to about 2.3.

You could try training the model again. If it still don't converge, please provide some more information, and maybe we can figure out the problem together.

**Comment by ahhan02** [comment]
Aha, I recompiled the project without cudnnv4, and it works now.... 

Anyway, I appreciate your help. @lhmRyan

**Comment by lhmRyan** [comment]
That's strange. I'll have a look at my code.

**Comment by liuyuying0829** [comment]
hello,anthor,nowadays,i read your cvpr2016 again,and i have a question about the loss in the paper,is the Eqn.(6),this examination said sigma(x)=1 when x alone to (-1,0)&[1,INF) and otherwise equal to -1, and Eqn.(5) use it, but in Eqn.(5),b(i,j) is a vector ? or this b(i,j) has been relaxed to a continuous variable ?

**Comment by liuyuying0829** [comment]
hello,anthor,nowadays,i read your cvpr2016 paper again,and i have a question about the loss in the paper,is the Eqn.(6), sigma(x)=1 when x alone to (-1,0)&[1,INF) and otherwise equal to -1, and Eqn.(5) use it, but in Eqn.(5),b(i,j) is a vector ? or this b(i,j) has been relaxed to a continuous variable ?

**Comment by lhmRyan** [comment]
@liuyuying0829 Hi. In Eqn(3)~(5), b_i,j have been relaxed to continuous vectors.

As explained below Eqn(6), this equation is applied to Eqn(5) element-wisely. Namely, Eqn(6) is applied to each element of b_i,j separately.
